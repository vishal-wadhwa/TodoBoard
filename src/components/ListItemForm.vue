<template>
  <v-card class='ma-1'>
    <v-card-title class='title'>{{formHeading}}</v-card-title>
    <v-divider></v-divider>
    <v-card-text class='pb-0'>
      <v-form ref='form'>
        <slot name='form'>
          <v-text-field
            autofocus
            v-model='title'
            prepend-icon='title'
            box
            :rules='[v => v && !!v.length || "Required"]'
            clear-icon='close'
            clearable
            label='Title'
          ></v-text-field>
          <v-textarea box v-model='desc' prepend-icon='description' rows='3'>
            <div slot='label'>
              Description
              <small>(optional)</small>
            </div>
          </v-textarea>
          <v-combobox
            box
            small-chips
            v-model='tagSelect'
            prepend-icon='category'
            deletable-chips
            multiple
            :chips='!!tagSelect.length'
            hint='Add a tag and press enter.'
            :items='tags'
            clearable
          >
            <div slot='label'>
              Tags
              <small>(optional)</small>
            </div>
          </v-combobox>
        </slot>
      </v-form>
    </v-card-text>
    <v-card-actions class='pt-0 pb-3'>
      <slot name='actions'>
        <v-layout justify-end>
          <v-btn round @click='discard' color='error'>
            <v-icon>clear</v-icon>&nbsp;Discard
          </v-btn>
          <v-btn round @click='save' color='success'>
            <v-icon>check_circle_outline</v-icon>&nbsp;Save
          </v-btn>
        </v-layout>
      </slot>
    </v-card-actions>
  </v-card>
</template>
<script>
// import BaseListItem from './BaseList/BaseListItem'

export default {
  data () {
    return {
      title: '',
      tagSelect: [],
      desc: ''
    }
  },
  props: {
    tags: {
      type: Array,
      default: () => []
    },
    formHeading: {
      type: String,
      default: 'Create new note'
    }
  },
  methods: {
    discard (ev) {
      this.$emit('lif:discard', ev)
      this.reset()
    },
    save (ev) {
      if (!this.$refs['form'].validate()) return
      this.$emit('lif:save', {
        title: this.title,
        tags: this.tagSelect,
        desc: this.desc
      }, ev)
      this.reset()
    },
    reset () {
      this.title = ''
      this.tagSelect = []
      this.desc = ''
      this.$refs['form'].reset()
    }
  }/* ,
  created () {
    console.log(BaseListItem.data())
  } */
}
</script>
<style lang="stylus" scoped>
</style>
