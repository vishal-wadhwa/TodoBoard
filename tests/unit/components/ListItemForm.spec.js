import { shallowMount, createLocalVue, mount } from '@vue/test-utils'
import ListItemForm from '@/components/ListItemForm'
import Vuetify from 'vuetify'
import Vue from 'vue'
Vue.use(Vuetify)

describe('ListItemForm.vue', () => {
  const app = document.createElement('div')
  app.setAttribute('data-app', true)
  document.body.appendChild(app)

  let localVue = null
  const factory = (opts = {}, full = false) => {
    return full ? mount(ListItemForm, opts) : shallowMount(ListItemForm, opts)
  }
  beforeEach(() => {
    localVue = createLocalVue()
    // localVue.use(Vuetify)
  })

  const tags = ['Categ1', 'Categ2']
  const formHeading = 'Add new entry'

  it('renders default form heading', () => {
    const wrapper = factory({
      localVue
    })
    expect(wrapper.text()).toMatch(ListItemForm.props.formHeading.default)
  })

  it('renders form heading from props', () => {
    const wrapper = factory({
      propsData: { formHeading },
      localVue
    })
    expect(wrapper.text()).not.toMatch(ListItemForm.props.formHeading.default)
    expect(wrapper.text()).toMatch(formHeading)
  })

  it('renders tags', () => {
    const wrapper = factory({
      stubs: ['v-combobox'],
      propsData: { tags },
      localVue
    })
    const comboboxWrapper = wrapper.find('v-combobox-stub')
    expect(comboboxWrapper.attributes('items')).toBe(tags.join(','))
  })

  it('emits "lif:discard" when discard button is clicked', () => {
    const wrapper = factory({
      localVue,
      sync: false
    }, true)

    const discardBtn = wrapper.find('button.error')
    discardBtn.trigger('click')

    const emitObj = wrapper.emitted('lif:discard')
    expect(emitObj).toBeTruthy()
  })

  it('does not emit "lif:save" on clicking when there is no title', () => {
    const wrapper = factory({
      localVue,
      sync: false
    }, true)

    const saveBtn = wrapper.find('button.success')
    saveBtn.trigger('click')

    const emitObj = wrapper.emitted('lif:save')
    expect(emitObj).toBeUndefined()
  })

  it('does not emit "lif:save" when title is empty or space-only string', () => {
    const wrapper = factory({
      localVue,
      sync: false
    }, true)

    wrapper.setData({ title: '  ' })
    const saveBtn = wrapper.find('button.success')
    saveBtn.trigger('click')

    const emitObj = wrapper.emitted('lif:save')
    expect(emitObj).toBeUndefined()
  })

  it('does not emit "lif:save" when one of the tags is empty or space-only string', () => {
    const wrapper = factory({
      localVue,
      sync: false
    }, true)

    wrapper.setData({ title: 'Note1', tagSelect: ['A', '  '] })
    const saveBtn = wrapper.find('button.success')
    saveBtn.trigger('click')

    const emitObj = wrapper.emitted('lif:save')
    expect(emitObj).toBeUndefined()
  })

  it('emits trimmed strings', () => {
    const wrapper = factory({
      localVue,
      stubs: ['v-text-field'], // suppresses stub issues: https://github.com/vuejs/vue-test-utils/issues/890
      sync: false
    }, true)
    const data = {
      title: 'Note1',
      tags: tags,
      desc: 'CDE'
    }
    wrapper.setData({
      title: '  ' + data['title'] + '   ',
      tagSelect: data['tags'].map(v => '  ' + v + '\n'),
      desc: data['desc'] + '\t\n\v'
    })

    const saveBtn = wrapper.find('button.success')
    saveBtn.trigger('click')

    const emitObj = wrapper.emitted('lif:save')[0][0]
    expect(emitObj).toEqual(data)
  })

  it('emits "lif:save" on clicking save button', () => {
    const wrapper = factory({
      localVue,
      stubs: ['v-text-field'], // suppresses stub issues: https://github.com/vuejs/vue-test-utils/issues/890
      sync: false
    }, true)
    const data = {
      title: 'Note1',
      tags: [],
      desc: ''
    }
    wrapper.setData({ title: data['title'] })
    const saveBtn = wrapper.find('button.success')
    saveBtn.trigger('click')

    const emitObj = wrapper.emitted('lif:save')[0][0]
    expect(emitObj).toEqual(data)
  })

  it('should match snapshot', () => {
    const wrapper = factory({
      propsData: { tags, formHeading },
      sync: false
    }, true)

    expect(wrapper.html()).toMatchSnapshot()
  })
})
