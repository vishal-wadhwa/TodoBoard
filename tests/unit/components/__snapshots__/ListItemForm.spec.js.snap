// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListItemForm.vue should match snapshot 1`] = `
<div class="ma-1 v-card theme--light">
  <div class="v-card__title title">Add new entry</div>
  <hr class="v-divider theme--light">
  <div class="v-card__text pb-0">
    <form novalidate="novalidate" class="v-form">
      <div class="v-input v-text-field v-text-field--box v-text-field--enclosed theme--light">
        <div class="v-input__prepend-outer">
          <div class="v-input__icon v-input__icon--prepend"><i aria-hidden="true" class="v-icon material-icons theme--light">title</i></div>
        </div>
        <div class="v-input__control">
          <div class="v-input__slot">
            <div class="v-text-field__slot"><label aria-hidden="true" class="v-label theme--light" style="left: 0px; position: absolute;">Title</label><input aria-label="Title" autofocus="autofocus" type="text"></div>
            <div class="v-input__append-inner">
              <div class="v-input__icon v-input__icon--"><i aria-hidden="true" class="v-icon v-icon--link material-icons theme--light"></i></div>
            </div>
          </div>
          <div class="v-text-field__details">
            <div class="v-messages theme--light">
              <div class="v-messages__wrapper"></div>
            </div>
          </div>
        </div>
      </div>
      <div class="v-input v-textarea v-text-field v-text-field--box v-text-field--enclosed theme--light">
        <div class="v-input__prepend-outer">
          <div class="v-input__icon v-input__icon--prepend"><i aria-hidden="true" class="v-icon material-icons theme--light">description</i></div>
        </div>
        <div class="v-input__control">
          <div class="v-input__slot">
            <div class="v-text-field__slot"><label aria-hidden="true" class="v-label theme--light" style="left: 0px; position: absolute;">
                <div>
                  Description
                  <small>(optional)</small></div>
              </label><textarea rows="3"></textarea></div>
          </div>
          <div class="v-text-field__details">
            <div class="v-messages theme--light">
              <div class="v-messages__wrapper"></div>
            </div>
          </div>
        </div>
      </div>
      <div class="v-input v-text-field v-text-field--box v-text-field--enclosed v-select v-select--chips v-select--chips--small v-autocomplete theme--light">
        <div class="v-input__prepend-outer">
          <div class="v-input__icon v-input__icon--prepend"><i aria-hidden="true" class="v-icon material-icons theme--light">category</i></div>
        </div>
        <div class="v-input__control">
          <div class="v-input__slot">
            <div class="v-select__slot"><label aria-hidden="true" class="v-label theme--light" style="left: 0px; position: absolute;">
                <div>
                  Tags
                  <small>(optional)</small></div>
              </label>
              <div class="v-select__selections"><input type="text" autocomplete="off" role="combobox"></div>
              <div class="v-input__append-inner">
                <div class="v-input__icon v-input__icon--"><i aria-hidden="true" class="v-icon v-icon--link material-icons theme--light"></i></div>
              </div>
              <div class="v-input__append-inner">
                <div class="v-input__icon v-input__icon--append"><i aria-hidden="true" class="v-icon material-icons theme--light">arrow_drop_down</i></div>
              </div>
            </div>
            <div class="v-menu"></div>
          </div>
          <div class="v-text-field__details">
            <div class="v-messages theme--light">
              <div class="v-messages__wrapper"></div>
            </div>
          </div>
        </div>
      </div>
    </form>
  </div>
  <div class="v-card__actions pt-0 pb-3">
    <div class="layout justify-end"><button type="button" class="v-btn v-btn--round theme--light error">
        <div class="v-btn__content"><i aria-hidden="true" class="v-icon material-icons theme--light">clear</i>&nbsp;Discard
        </div>
      </button> <button type="button" class="v-btn v-btn--round theme--light success">
        <div class="v-btn__content"><i aria-hidden="true" class="v-icon material-icons theme--light">check_circle_outline</i>&nbsp;Save
        </div>
      </button></div>
  </div>
</div>
`;
